diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1c4cfdd..74b2b28 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -369,6 +369,11 @@ add_feature_info(libdisplayinfo libdisplayinfo_FOUND "EDID and DisplayID library
 
 pkg_check_modules(PipeWire IMPORTED_TARGET libpipewire-0.3>=1.0.9)
 add_feature_info(PipeWire PipeWire_FOUND "Required for Wayland screencasting")
+if (PipeWire_FOUND AND ${PipeWire_VERSION} VERSION_GREATER_EQUAL 1.5.81)
+    set(HAVE_PIPEWIRE_ST2084 1)
+else()
+    set(HAVE_PIPEWIRE_ST2084 0)
+endif()
 
 if (KWIN_BUILD_NOTIFICATIONS)
     find_package(KF6 ${KF6_MIN_VERSION} REQUIRED COMPONENTS Notifications)
diff --git a/src/config-kwin.h.cmake b/src/config-kwin.h.cmake
index 6c652f1..80689ff 100644
--- a/src/config-kwin.h.cmake
+++ b/src/config-kwin.h.cmake
@@ -28,6 +28,7 @@ constexpr QLatin1String LIBEXEC_DIR("${CMAKE_INSTALL_FULL_LIBEXECDIR}");
 #cmakedefine01 HAVE_XWAYLAND_ENABLE_EI_PORTAL
 #cmakedefine01 HAVE_DL_LIBRARY
 #cmakedefine01 HAVE_LIBDRM_FAUX
+#cmakedefine01 HAVE_PIPEWIRE_ST2084
 
 constexpr QLatin1String XWAYLAND_SESSION_SCRIPTS("${XWAYLAND_SESSION_SCRIPTS}");
 #cmakedefine01 HAVE_WAYLAND_PROTOCOLS_147
diff --git a/src/plugins/screencast/outputscreencastsource.cpp b/src/plugins/screencast/outputscreencastsource.cpp
index 0fd74bf..4d9b3bd 100644
--- a/src/plugins/screencast/outputscreencastsource.cpp
+++ b/src/plugins/screencast/outputscreencastsource.cpp
@@ -46,11 +46,6 @@ OutputScreenCastSource::~OutputScreenCastSource()
     pause();
 }
 
-quint32 OutputScreenCastSource::drmFormat() const
-{
-    return DRM_FORMAT_ARGB8888;
-}
-
 QSize OutputScreenCastSource::textureSize() const
 {
     return m_output->pixelSize();
@@ -61,6 +56,11 @@ qreal OutputScreenCastSource::devicePixelRatio() const
     return m_output->scale();
 }
 
+bool OutputScreenCastSource::needsAlpha() const
+{
+    return false;
+}
+
 void OutputScreenCastSource::setRenderCursor(bool enable)
 {
     m_renderCursor = enable;
@@ -69,6 +69,12 @@ void OutputScreenCastSource::setRenderCursor(bool enable)
     }
 }
 
+void OutputScreenCastSource::setColorDescription(const std::shared_ptr<ColorDescription> &color)
+{
+    Q_ASSERT(m_layer);
+    m_layer->setColorDescription(color);
+}
+
 Region OutputScreenCastSource::render(QImage *target, const Region &bufferRepair)
 {
     auto texture = GLTexture::allocate(GL_RGBA8, target->size());
diff --git a/src/plugins/screencast/outputscreencastsource.h b/src/plugins/screencast/outputscreencastsource.h
index 7c3142b..6c2e2c9 100644
--- a/src/plugins/screencast/outputscreencastsource.h
+++ b/src/plugins/screencast/outputscreencastsource.h
@@ -29,9 +29,10 @@ public:
     uint refreshRate() const override;
     QSize textureSize() const override;
     qreal devicePixelRatio() const override;
-    quint32 drmFormat() const override;
+    bool needsAlpha() const override;
 
     void setRenderCursor(bool enable) override;
+    void setColorDescription(const std::shared_ptr<ColorDescription> &color) override;
     Region render(GLFramebuffer *target, const Region &bufferRepair) override;
     Region render(QImage *target, const Region &bufferRepair) override;
     std::chrono::nanoseconds clock() const override;
diff --git a/src/plugins/screencast/regionscreencastsource.cpp b/src/plugins/screencast/regionscreencastsource.cpp
index 8e0ff9c..20451d5 100644
--- a/src/plugins/screencast/regionscreencastsource.cpp
+++ b/src/plugins/screencast/regionscreencastsource.cpp
@@ -52,14 +52,14 @@ qreal RegionScreenCastSource::devicePixelRatio() const
     return m_scale;
 }
 
-quint32 RegionScreenCastSource::drmFormat() const
+std::chrono::nanoseconds RegionScreenCastSource::clock() const
 {
-    return DRM_FORMAT_ARGB8888;
+    return m_last;
 }
 
-std::chrono::nanoseconds RegionScreenCastSource::clock() const
+bool RegionScreenCastSource::needsAlpha() const
 {
-    return m_last;
+    return false;
 }
 
 void RegionScreenCastSource::setRenderCursor(bool enable)
@@ -70,6 +70,12 @@ void RegionScreenCastSource::setRenderCursor(bool enable)
     }
 }
 
+void RegionScreenCastSource::setColorDescription(const std::shared_ptr<ColorDescription> &color)
+{
+    Q_ASSERT(m_layer);
+    m_layer->setColorDescription(color);
+}
+
 Region RegionScreenCastSource::render(GLFramebuffer *target, const Region &bufferRepair)
 {
     m_last = std::chrono::steady_clock::now().time_since_epoch();
diff --git a/src/plugins/screencast/regionscreencastsource.h b/src/plugins/screencast/regionscreencastsource.h
index b9fd361..625276a 100644
--- a/src/plugins/screencast/regionscreencastsource.h
+++ b/src/plugins/screencast/regionscreencastsource.h
@@ -25,12 +25,13 @@ public:
     explicit RegionScreenCastSource(const Rect &region, qreal scale, std::optional<pid_t> pidToHide);
     ~RegionScreenCastSource() override;
 
-    quint32 drmFormat() const override;
     QSize textureSize() const override;
     qreal devicePixelRatio() const override;
     uint refreshRate() const override;
+    bool needsAlpha() const override;
 
     void setRenderCursor(bool enable) override;
+    void setColorDescription(const std::shared_ptr<ColorDescription> &color) override;
     Region render(GLFramebuffer *target, const Region &bufferRepair) override;
     Region render(QImage *target, const Region &bufferRepair) override;
     std::chrono::nanoseconds clock() const override;
diff --git a/src/plugins/screencast/screencastlayer.cpp b/src/plugins/screencast/screencastlayer.cpp
index 62f7986..fe1c50e 100644
--- a/src/plugins/screencast/screencastlayer.cpp
+++ b/src/plugins/screencast/screencastlayer.cpp
@@ -16,6 +16,15 @@ ScreencastLayer::ScreencastLayer(LogicalOutput *output, const QHash<uint32_t, QL
     setRenderLoop(nullptr);
 }
 
+void ScreencastLayer::setColorDescription(const std::shared_ptr<ColorDescription> &color)
+{
+    if (color == m_color) {
+        return;
+    }
+    m_color = color;
+    addDeviceRepaint(Rect(QPoint(), targetRect().size()));
+}
+
 void ScreencastLayer::setFramebuffer(GLFramebuffer *buffer, const Region &bufferDamage)
 {
     // TODO is there a better way to deal with this?
@@ -36,7 +45,7 @@ QHash<uint32_t, QList<uint64_t>> ScreencastLayer::supportedDrmFormats() const
 std::optional<OutputLayerBeginFrameInfo> ScreencastLayer::doBeginFrame()
 {
     return OutputLayerBeginFrameInfo{
-        .renderTarget = RenderTarget(m_buffer),
+        .renderTarget = RenderTarget(m_buffer, m_color),
         .repaint = m_bufferDamage,
     };
 }
diff --git a/src/plugins/screencast/screencastlayer.h b/src/plugins/screencast/screencastlayer.h
index f9c0f2e..41c5faa 100644
--- a/src/plugins/screencast/screencastlayer.h
+++ b/src/plugins/screencast/screencastlayer.h
@@ -4,6 +4,7 @@
     SPDX-License-Identifier: GPL-2.0-or-later
 */
 #pragma once
+#include "core/colorspace.h"
 #include "core/outputlayer.h"
 
 namespace KWin
@@ -14,6 +15,7 @@ class ScreencastLayer : public OutputLayer
 public:
     explicit ScreencastLayer(LogicalOutput *output, const QHash<uint32_t, QList<uint64_t>> &formats);
 
+    void setColorDescription(const std::shared_ptr<ColorDescription> &color);
     void setFramebuffer(GLFramebuffer *buffer, const Region &bufferDamage);
 
     DrmDevice *scanoutDevice() const override;
@@ -27,6 +29,7 @@ private:
     const QHash<uint32_t, QList<uint64_t>> m_formats;
     GLFramebuffer *m_buffer = nullptr;
     Region m_bufferDamage;
+    std::shared_ptr<ColorDescription> m_color = ColorDescription::sRGB;
 };
 
 }
diff --git a/src/plugins/screencast/screencastsource.h b/src/plugins/screencast/screencastsource.h
index c15f812..4173ce7 100644
--- a/src/plugins/screencast/screencastsource.h
+++ b/src/plugins/screencast/screencastsource.h
@@ -7,6 +7,7 @@
 #pragma once
 
 #include <QObject>
+#include <memory>
 
 class QImage;
 
@@ -18,6 +19,7 @@ class GLFramebuffer;
 class GLTexture;
 class RectF;
 class Region;
+class ColorDescription;
 
 class ScreenCastSource : public QObject
 {
@@ -27,11 +29,12 @@ public:
     explicit ScreenCastSource();
 
     virtual uint refreshRate() const = 0;
-    virtual quint32 drmFormat() const = 0;
     virtual QSize textureSize() const = 0;
     virtual qreal devicePixelRatio() const = 0;
+    virtual bool needsAlpha() const = 0;
 
     virtual void setRenderCursor(bool enable) = 0;
+    virtual void setColorDescription(const std::shared_ptr<ColorDescription> &color) = 0;
     virtual Region render(GLFramebuffer *target, const Region &bufferRepair) = 0;
     virtual Region render(QImage *target, const Region &bufferRepair) = 0;
     virtual std::chrono::nanoseconds clock() const = 0;
diff --git a/src/plugins/screencast/screencaststream.cpp b/src/plugins/screencast/screencaststream.cpp
index 87584fc..53f36fa 100644
--- a/src/plugins/screencast/screencaststream.cpp
+++ b/src/plugins/screencast/screencaststream.cpp
@@ -8,6 +8,7 @@
 
 #include "screencaststream.h"
 #include "compositor.h"
+#include "config-kwin.h"
 #include "core/drmdevice.h"
 #include "core/graphicsbufferallocator.h"
 #include "core/renderbackend.h"
@@ -24,6 +25,9 @@
 #include "screencastsource.h"
 #include "utils/drm_format_helper.h"
 
+#include <algorithm>
+#include <span>
+
 #include <KLocalizedString>
 
 #include <QLoggingCategory>
@@ -73,10 +77,6 @@ static const struct
         .drmFormat = DRM_FORMAT_BGRX8888,
         .spaFormat = SPA_VIDEO_FORMAT_xRGB,
     },
-    {
-        .drmFormat = DRM_FORMAT_NV12,
-        .spaFormat = SPA_VIDEO_FORMAT_NV12,
-    },
     {
         .drmFormat = DRM_FORMAT_RGB888,
         .spaFormat = SPA_VIDEO_FORMAT_BGR,
@@ -85,6 +85,35 @@ static const struct
         .drmFormat = DRM_FORMAT_BGR888,
         .spaFormat = SPA_VIDEO_FORMAT_RGB,
     },
+    {
+        // No alpha. Preferred by Sunshine's HDR path (gates BT2020+SMPTE2084 on this format).
+        .drmFormat = DRM_FORMAT_XBGR2101010,
+        .spaFormat = SPA_VIDEO_FORMAT_xBGR_210LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_XRGB2101010,
+        .spaFormat = SPA_VIDEO_FORMAT_xRGB_210LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_BGRA1010102,
+        .spaFormat = SPA_VIDEO_FORMAT_ARGB_210LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_RGBA1010102,
+        .spaFormat = SPA_VIDEO_FORMAT_ABGR_210LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_ABGR2101010,
+        .spaFormat = SPA_VIDEO_FORMAT_RGBA_102LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_ARGB2101010,
+        .spaFormat = SPA_VIDEO_FORMAT_BGRA_102LE,
+    },
+    {
+        .drmFormat = DRM_FORMAT_ABGR16161616F,
+        .spaFormat = SPA_VIDEO_FORMAT_RGBA_F16,
+    },
 };
 
 static spa_video_format drmFormatToSpaVideoFormat(quint32 drmFormat)
@@ -151,8 +180,56 @@ void ScreenCastStream::newStreamParams()
 {
     qCDebug(KWIN_SCREENCAST) << objectName() << "announcing stream params. with dmabuf:" << m_dmabufParams.has_value();
     const int buffertypes = m_dmabufParams ? (1 << SPA_DATA_DmaBuf) : (1 << SPA_DATA_MemFd);
-    const int bpp = m_videoFormat.format == SPA_VIDEO_FORMAT_RGB || m_videoFormat.format == SPA_VIDEO_FORMAT_BGR ? 3 : 4;
-    const int stride = SPA_ROUND_UP_N(m_resolution.width() * bpp, 4);
+
+    Colorimetry primaries = Colorimetry::BT709;
+    switch (m_videoFormat.color_primaries) {
+    case SPA_VIDEO_COLOR_PRIMARIES_UNKNOWN:
+    case SPA_VIDEO_COLOR_PRIMARIES_BT709:
+        primaries = Colorimetry::BT709;
+        break;
+    case SPA_VIDEO_COLOR_PRIMARIES_BT2020:
+        primaries = Colorimetry::BT2020;
+        break;
+    case SPA_VIDEO_COLOR_PRIMARIES_ADOBERGB:
+        primaries = Colorimetry::AdobeRGB;
+        break;
+    case SPA_VIDEO_COLOR_PRIMARIES_BT470BG:
+        primaries = Colorimetry::PAL;
+        break;
+    case SPA_VIDEO_COLOR_PRIMARIES_BT470M:
+        primaries = Colorimetry::PAL_M;
+        break;
+        // not sure why Pipewire has these as separate entries,
+        // they're the same actual primaries
+    case SPA_VIDEO_COLOR_PRIMARIES_SMPTE170M:
+    case SPA_VIDEO_COLOR_PRIMARIES_SMPTE240M:
+        primaries = Colorimetry::NTSC;
+        break;
+    case SPA_VIDEO_COLOR_PRIMARIES_FILM:
+        primaries = Colorimetry::GenericFilm;
+        break;
+    default:
+        break;
+    }
+    TransferFunction::Type transfer = TransferFunction::gamma22;
+    switch (m_videoFormat.transfer_function) {
+    case SPA_VIDEO_TRANSFER_BT601:
+    case SPA_VIDEO_TRANSFER_BT709:
+    case SPA_VIDEO_TRANSFER_BT2020_10:
+        transfer = TransferFunction::BT1886;
+        break;
+#if HAVE_PIPEWIRE_ST2084
+    case SPA_VIDEO_TRANSFER_SMPTE2084:
+        transfer = TransferFunction::PerceptualQuantizer;
+        break;
+#endif
+    case SPA_VIDEO_TRANSFER_GAMMA22:
+    default:
+        break;
+    }
+    m_videoColor = std::make_shared<ColorDescription>(primaries, TransferFunction{transfer});
+
+    qCDebug(KWIN_SCREENCAST) << "chose color description for the stream";
 
     struct spa_pod_dynamic_builder pod_builder;
     struct spa_pod_frame f;
@@ -179,6 +256,8 @@ void ScreenCastStream::newStreamParams()
                         SPA_PARAM_BUFFERS_buffers, SPA_POD_CHOICE_RANGE_Int(3, 2, 4),
                         SPA_PARAM_BUFFERS_dataType, SPA_POD_CHOICE_FLAGS_Int(buffertypes), 0);
     if (!m_dmabufParams) {
+        const int bpp = m_videoFormat.format == SPA_VIDEO_FORMAT_RGB || m_videoFormat.format == SPA_VIDEO_FORMAT_BGR ? 3 : 4;
+        const int stride = SPA_ROUND_UP_N(m_resolution.width() * bpp, 4);
         spa_pod_builder_add(&pod_builder.b,
                             SPA_PARAM_BUFFERS_blocks, SPA_POD_Int(1),
                             SPA_PARAM_BUFFERS_size, SPA_POD_Int(stride * m_resolution.height()),
@@ -229,8 +308,8 @@ void ScreenCastStream::onStreamParamChanged(uint32_t id, const struct spa_pod *f
     }
 
     spa_format_video_raw_parse(format, &m_videoFormat);
-    auto modifierProperty = spa_pod_find_prop(format, nullptr, SPA_FORMAT_VIDEO_modifier);
-    if (modifierProperty) {
+    const bool reallocationRequired = !m_dmabufParams || m_dmabufParams->width != m_resolution.width() || m_dmabufParams->height != m_resolution.height();
+    if (auto modifierProperty = spa_pod_find_prop(format, nullptr, SPA_FORMAT_VIDEO_modifier)) {
         const uint32_t valueCount = SPA_POD_CHOICE_N_VALUES(&modifierProperty->value);
         const uint64_t *values = (uint64_t *)SPA_POD_CHOICE_VALUES(&modifierProperty->value); // values[0] is the preferred choice
 
@@ -244,7 +323,7 @@ void ScreenCastStream::onStreamParamChanged(uint32_t id, const struct spa_pod *f
             }
         }
 
-        if (!m_dmabufParams || m_dmabufParams->width != m_resolution.width() || m_dmabufParams->height != m_resolution.height() || !receivedModifiers.contains(m_dmabufParams->modifier)) {
+        if (reallocationRequired || !receivedModifiers.contains(m_dmabufParams->modifier)) {
             // DRM_MOD_INVALID should be used as a last option. Do not just remove it it's the only
             // item on the list
             if (receivedModifiers.count() > 1) {
@@ -261,12 +340,51 @@ void ScreenCastStream::onStreamParamChanged(uint32_t id, const struct spa_pod *f
                 }
             }
 
-            qCDebug(KWIN_SCREENCAST) << objectName() << "Stream dmabuf modifiers received, offering our best suited modifier" << m_dmabufParams.has_value();
+            qCWarning(KWIN_SCREENCAST) << objectName() << "Stream dmabuf modifiers received, offering our best suited modifier" << m_dmabufParams.has_value();
             char buffer[2048];
             auto params = buildFormats(m_dmabufParams.has_value(), buffer);
             pw_stream_update_params(m_pwStream, params.data(), params.count());
             return;
         }
+    } else if (auto formatProperty = spa_pod_find_prop(format, nullptr, SPA_FORMAT_VIDEO_format)) {
+        const uint32_t valueCount = SPA_POD_CHOICE_N_VALUES(&formatProperty->value);
+        const spa_video_format *values = (spa_video_format *)SPA_POD_CHOICE_VALUES(&formatProperty->value);
+        bool currentFormatOffered = false;
+        for (uint32_t i = 0; i < valueCount; ++i) {
+            if (static_cast<uint32_t>(values[i]) == m_videoFormat.format) {
+                currentFormatOffered = true;
+                break;
+            }
+        }
+        if (reallocationRequired || !currentFormatOffered) {
+            // attempt to change the format to something the client supports
+            const auto supported = Compositor::self()->backend()->supportedFormats();
+            for (uint32_t i = 0; i < valueCount; ++i) {
+                const uint32_t drmFormat = spaVideoFormatToDrmFormat(values[i]);
+                if (drmFormat == DRM_FORMAT_INVALID) {
+                    continue;
+                }
+                if (!m_possibleDmabufFormats.contains(drmFormat)) {
+                    continue;
+                }
+                const auto mods = supported.value(drmFormat);
+                const auto opt = testCreateDmaBuf(m_resolution, drmFormat, mods);
+                if (opt.has_value()) {
+                    m_dmabufParams = opt;
+                    m_drmFormat = drmFormat;
+                    m_modifiers = mods;
+
+                    qCDebug(KWIN_SCREENCAST) << objectName() << "Changing stream format to" << FormatInfo::drmFormatName(m_drmFormat);
+                    char buffer[2048];
+                    auto params = buildFormats(m_dmabufParams.has_value(), buffer);
+                    pw_stream_update_params(m_pwStream, params.data(), params.count());
+                    return;
+                } else {
+                    // avoid trying this again later on
+                    m_possibleDmabufFormats.removeOne(drmFormat);
+                }
+            }
+        }
     } else {
         m_dmabufParams.reset();
     }
@@ -401,29 +519,66 @@ uint ScreenCastStream::nodeId()
     return m_pwNodeId;
 }
 
-bool ScreenCastStream::createStream()
+static QList<uint32_t> filterAndSort(const QHash<uint32_t, QList<uint64_t>> &supported, bool needsAlpha)
 {
-    const QByteArray objname = "kwin-screencast-" + objectName().toUtf8();
-    m_pwStream = pw_stream_new(m_pwCore->pwCore, objname, nullptr);
+    QList<uint32_t> ret = supported.keys();
+    // remove all formats we don't want or can't support
+    ret.erase(std::remove_if(ret.begin(), ret.end(), [needsAlpha](uint32_t format) {
+                  bool inList = false;
+                  for (const auto &pair : supportedFormats) {
+                      if (pair.drmFormat == format) {
+                          inList = true;
+                          break;
+                      }
+                  }
+                  if (!inList) {
+                      return true;
+                  }
+                  const auto info = FormatInfo::get(format);
+                  if (!info.has_value()) {
+                      return true;
+                  }
+                  return info->bitsPerColor < 8 || info->bitsPerPixel > 64 || (needsAlpha && info->alphaBits < 8);
+              }),
+              ret.end());
+    std::sort(ret.begin(), ret.end(), [needsAlpha](uint32_t left, uint32_t right) {
+        const auto leftInfo = FormatInfo::get(left);
+        const auto rightInfo = FormatInfo::get(right);
+        // Prefer 10bpc if possible, as it's both high bit depth and low bandwidth
+        const bool left10 = leftInfo->bitsPerColor == 10;
+        const bool right10 = rightInfo->bitsPerColor == 10;
+        if (left10 != right10) {
+            return left10;
+        }
+        // Prefer higher bit depths over more efficiency.
+        if (leftInfo->bitsPerColor != rightInfo->bitsPerColor) {
+            return leftInfo->bitsPerColor > rightInfo->bitsPerColor;
+        }
+        // When alpha isn't needed, prefer alpha-less variants. This lets HDR-aware
+        // clients (e.g. Sunshine) that gate HDR colorspace on xBGR_210LE specifically
+        // see their preferred format negotiated.
+        if (!needsAlpha && leftInfo->alphaBits != rightInfo->alphaBits) {
+            return leftInfo->alphaBits < rightInfo->alphaBits;
+        }
+        // At the same bit depth, prefer low bandwidth for better performance
+        return leftInfo->bitsPerPixel < rightInfo->bitsPerPixel;
+    });
+    return ret;
+}
 
+bool ScreenCastStream::createStream()
+{
     const auto supported = Compositor::self()->backend()->supportedFormats();
-    auto itModifiers = supported.constFind(m_source->drmFormat());
-
-    // If the offered format is not available for dmabuf, prefer converting to another one than resorting to memfd
-    if (itModifiers == supported.constEnd() && !supported.isEmpty()) {
-        itModifiers = supported.constFind(DRM_FORMAT_ARGB8888);
-        if (itModifiers != supported.constEnd()) {
-            m_drmFormat = itModifiers.key();
-        }
+    m_possibleDmabufFormats = filterAndSort(supported, m_source->needsAlpha());
+    if (m_possibleDmabufFormats.isEmpty()) {
+        return false;
     }
 
-    if (itModifiers == supported.constEnd()) {
-        m_drmFormat = m_source->drmFormat();
-        m_modifiers = {DRM_FORMAT_MOD_INVALID};
-    } else {
-        m_drmFormat = itModifiers.key();
-        m_modifiers = *itModifiers;
-    }
+    const QByteArray objname = "kwin-screencast-" + objectName().toUtf8();
+    m_pwStream = pw_stream_new(m_pwCore->pwCore, objname, nullptr);
+
+    m_drmFormat = m_possibleDmabufFormats.front();
+    m_modifiers = supported.value(m_drmFormat);
     m_hasDmaBuf = testCreateDmaBuf(m_resolution, m_drmFormat, m_modifiers).has_value();
 
     char buffer[2048];
@@ -601,6 +756,8 @@ void ScreenCastStream::record(Contents contents)
         break;
     }
 
+    m_source->setColorDescription(m_videoColor);
+
     EglContext *context = backend->openglContext();
     context->makeCurrent();
 
@@ -778,6 +935,56 @@ spa_pod *ScreenCastStream::buildFormat(struct spa_pod_builder *b, enum spa_video
                             SPA_POD_Fraction(maxFramerate)),
                         0);
 
+    // The default color primaries depend on the video format: for 10-bit
+    // formats we prefer BT2020 (HDR), otherwise BT709 (SDR).
+    struct spa_pod_frame primariesFrame;
+    spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_colorPrimaries, 0);
+    spa_pod_builder_push_choice(b, &primariesFrame, SPA_CHOICE_Enum, 0);
+    if (format == SPA_VIDEO_FORMAT_ARGB_210LE ||
+        format == SPA_VIDEO_FORMAT_ABGR_210LE ||
+        format == SPA_VIDEO_FORMAT_RGBA_102LE ||
+        format == SPA_VIDEO_FORMAT_BGRA_102LE) {
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT2020); // default
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT2020);
+    } else {
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT709); // default
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT709);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT2020);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_ADOBERGB);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT470M);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_BT470BG);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_SMPTE170M);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_SMPTE240M);
+        spa_pod_builder_id(b, SPA_VIDEO_COLOR_PRIMARIES_FILM);
+    }
+    spa_pod_builder_pop(b, &primariesFrame);
+
+    // The default transfer function depends on the video format: for 10-bit
+    // formats we prefer SMPTE2084 (PQ HDR), otherwise gamma22 (SDR).
+    struct spa_pod_frame transferFrame;
+    spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_transferFunction, 0);
+    spa_pod_builder_push_choice(b, &transferFrame, SPA_CHOICE_Enum, 0);
+#if HAVE_PIPEWIRE_ST2084
+    if (format == SPA_VIDEO_FORMAT_ARGB_210LE ||
+        format == SPA_VIDEO_FORMAT_ABGR_210LE ||
+        format == SPA_VIDEO_FORMAT_RGBA_102LE ||
+        format == SPA_VIDEO_FORMAT_BGRA_102LE) {
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_SMPTE2084); // default
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_SMPTE2084);
+    } else
+#endif
+    {
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_GAMMA22); // default
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_GAMMA22);
+#if HAVE_PIPEWIRE_ST2084
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_SMPTE2084);
+#endif
+        // SPA_VIDEO_TRANSFER_SRGB is intentionally not supported,
+        // as it would almost certainly yield wrong results.
+        spa_pod_builder_id(b, SPA_VIDEO_TRANSFER_BT709);
+    }
+    spa_pod_builder_pop(b, &transferFrame);
+
     if (format == SPA_VIDEO_FORMAT_BGRA) {
         /* announce equivalent format without alpha */
         spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_CHOICE_ENUM_Id(3, format, format, SPA_VIDEO_FORMAT_BGRx), 0);
diff --git a/src/plugins/screencast/screencaststream.h b/src/plugins/screencast/screencaststream.h
index 9518b58..21b50c9 100644
--- a/src/plugins/screencast/screencaststream.h
+++ b/src/plugins/screencast/screencaststream.h
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include "core/colorspace.h"
 #include "utils/damagejournal.h"
 #include "wayland/screencast_v1.h"
 
@@ -115,9 +116,11 @@ private:
     bool m_closed = false;
 
     spa_video_info_raw m_videoFormat;
+    std::shared_ptr<ColorDescription> m_videoColor = ColorDescription::sRGB;
     QString m_error;
     QList<uint64_t> m_modifiers;
     std::optional<ScreenCastDmaBufTextureParams> m_dmabufParams; // when fixated
+    QList<uint32_t> m_possibleDmabufFormats;
 
     struct
     {
diff --git a/src/plugins/screencast/windowscreencastsource.cpp b/src/plugins/screencast/windowscreencastsource.cpp
index 5970897..269a0bc 100644
--- a/src/plugins/screencast/windowscreencastsource.cpp
+++ b/src/plugins/screencast/windowscreencastsource.cpp
@@ -88,11 +88,6 @@ void WindowScreenCastSource::unwatch(Window *window)
     disconnect(window, &Window::damaged, this, &WindowScreenCastSource::frame);
 }
 
-quint32 WindowScreenCastSource::drmFormat() const
-{
-    return DRM_FORMAT_ARGB8888;
-}
-
 QSize WindowScreenCastSource::textureSize() const
 {
     return (boundingRect().size() * m_windows[0]->targetScale()).toSize();
@@ -103,11 +98,21 @@ qreal WindowScreenCastSource::devicePixelRatio() const
     return m_windows[0]->targetScale();
 }
 
+bool WindowScreenCastSource::needsAlpha() const
+{
+    return true;
+}
+
 void WindowScreenCastSource::setRenderCursor(bool enable)
 {
     m_renderCursor = enable;
 }
 
+void WindowScreenCastSource::setColorDescription(const std::shared_ptr<ColorDescription> &color)
+{
+    m_color = color;
+}
+
 Region WindowScreenCastSource::render(QImage *target, const Region &bufferDamage)
 {
     const auto offscreenTexture = GLTexture::allocate(GL_RGBA8, target->size());
@@ -124,7 +129,7 @@ Region WindowScreenCastSource::render(QImage *target, const Region &bufferDamage
 
 Region WindowScreenCastSource::render(GLFramebuffer *target, const Region &bufferDamage)
 {
-    RenderTarget renderTarget(target);
+    RenderTarget renderTarget(target, m_color);
     RenderViewport viewport(boundingRect(), devicePixelRatio(), renderTarget, QPoint());
 
     WorkspaceScene *scene = Compositor::self()->scene();
diff --git a/src/plugins/screencast/windowscreencastsource.h b/src/plugins/screencast/windowscreencastsource.h
index 80edd86..f301188 100644
--- a/src/plugins/screencast/windowscreencastsource.h
+++ b/src/plugins/screencast/windowscreencastsource.h
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include "core/colorspace.h"
 #include "screencastsource.h"
 #include "window.h"
 
@@ -23,12 +24,13 @@ public:
     explicit WindowScreenCastSource(Window *window);
     ~WindowScreenCastSource() override;
 
-    quint32 drmFormat() const override;
     QSize textureSize() const override;
     qreal devicePixelRatio() const override;
     uint refreshRate() const override;
+    bool needsAlpha() const override;
 
     void setRenderCursor(bool enable) override;
+    void setColorDescription(const std::shared_ptr<ColorDescription> &color) override;
     Region render(GLFramebuffer *target, const Region &bufferDamage) override;
     Region render(QImage *target, const Region &bufferDamage) override;
     std::chrono::nanoseconds clock() const override;
@@ -50,6 +52,7 @@ private:
     QList<Window *> m_windows;
     bool m_active = false;
     bool m_renderCursor = false;
+    std::shared_ptr<ColorDescription> m_color = ColorDescription::sRGB;
 };
 
 } // namespace KWin
